docs/docs/dev_docs/contracts/events.md

```diff
@@ -1,17 +1,104 @@
+
+At this point we only allow emitting note spending info through encrypted events.
+In the future we will allow emitting arbitrary information.
+(If you currently emit arbitrary information, AztecRPC server will fail to process it and it will not be queryable.)
+
+To emit encrypted logs first import the `emit_encrypted_log` utility function inside your contract:
+
+#include_code encrypted_import /yarn-project/noir-libs/value-note/src/utils.nr rust
+
+Then you can call the function:
+
+#include_code encrypted /yarn-project/noir-libs/value-note/src/utils.nr rust
 
-### Constraining events
 
 ### Unencrypted Events
+Unencrypted events are events which can be read by anyone.
+They can be emitted by both public and private functions.
+It's important for a developer to consider whether it's acceptable for the unencrypted event to be emitted from private functions as it might pose a significant privacy leak.
 
-### Encrypted Events
+Once emitted, unencrypted events are stored in AztecNode and can be queried by anyone:
+<Tabs groupId="events">
+<TabItem value="cli" label="Aztec CLI">
+
+```bash
+aztec-cli get-logs --from 5 --limit 1
+```
+
+</TabItem>
+<TabItem value="js" label="Aztec.js">
+
+#include_code logs /yarn-project/end-to-end/src/e2e_public_token_contract.test.ts typescript
+
+</TabItem>
+</Tabs>
+
+To emit unencrypted logs first import the `emit_unencrypted_log` utility function inside your contract:
+
+#include_code unencrypted_import /yarn-project/noir-contracts/src/contracts/private_token_contract/src/main.nr rust
+
+Then you can call the function:
+
+#include_code unencrypted /yarn-project/noir-contracts/src/contracts/private_token_contract/src/main.nr rust
 
 ### Costs
 
-Explain L1 cost to emit an event.
+All the event data is pushed on Ethereum by the sequencer and for this reason the cost of emitting an event is non-trivial.
 
 ## Processing events
+Both the encrypted and unencrypted events are stored in AztecNode.
+Unencrypted logs can be queried by anyone as we described above in the [Unencrypted Events](#unencrypted-events) section.
+
+Encrypted logs need to first be decrypted:
 
 ### Decrypting
+One function of AztecRPC server is constantly loading encrypted logs from AztecNode and trying to decrypt them.
+When new encrypted logs are obtained, AztecRPC server will try to decrypt them using the private encryption key of all the accounts registered inside AztecRPC server.
+If the decryption is successful, AztecRPC server will store the decrypted note inside a database.
+If the decryption fails, the specific log will be discarded.
+
+For the AztecRPC server to successfully process the decrypted note we need to compute note hash and nullifier.
+Because we want to support arbitrary ways of computing these values, Noir developers need to implement a `compute_note_hash_and_nullifier` function inside their contracts.
```

Comment:
```suggestion
Aztec.nr enables smart contract developers to design custom notes, meaning developers can also customise how a note's note hash and nullifier are to be computed. Because of this customisability, and because there will be a potentially-unlimited number of smart contracts deployed to Aztec, all Aztec nodes need to be 'taught' how to compute the custom note hashes and nullifiers for a particular contract. Therefore, developers will need to implement a `compute_note_hash_and_nullifier` function inside their contracts.
```